### ISSUE TITLE
Interface should not have a constructor

### ISSUE THREAD

**Issue by SamMousa** [yiisoft/di#54]
Currently `ServiceProviderInterface` has 2 functions:
```php
    public function __construct(ContainerInterface $container);
    public function register(): void;

```

Interfaces generally should not have constructors. 
In this case I propose changing it to this:
```php
public function register(ContainerInterface $container): void;
```

This has several advantages:
- I could create the container after the service provider.
- I could serialize my service provider.

I don't have any need for these advantages, but I think the code is cleaner and more correct this way.

**Comment by samdark** [comment]
> I could create the container after the service provider.

Why is this an advantage?

**Comment by SamMousa** [comment]
Not sure, that's why I said I don't have any need for them ;-)

**Comment by samdark** [comment]
Another question. Does ServiceProvider make any sense without a Container registered?

**Comment by SamMousa** [comment]
Yeah, you could have one serialized or use one for multiple containers.

It doesn't need to store a reference to a container to do its job.

**Comment by SamMousa** [comment]
Implemented in #57
